CoreCLR support for default interface implementation of static virtual methods

trylek opened the pull request:
This change introduces a "static virtual" version of the diamondshape test and makes
runtime changes to make it functional. As default interface implementations can cause
ambiguities in the presence of diamond pattern (when a class implements two interfaces
and each of them contains a default interface implementation of a static virtual method
defined in an interface they both inherit from), the gist of the runtime / JIT change deals
with this ambiguity that wasn't seen previously in static virtual methods without default
interface implementations.

When the static virtual method resolution in TryResolveConstraintMethodApprox called
from getCallInfo fail due to ambiguity, we take further action based on whether the
SVM call involves shared generics. In the presence of shared generics we cannot always
determine the SVM call target at compile time so we use the ConstrainedMethodEntrySlot
dictionary entry to represent the runtime resolution. In the absence of shared generics,
we instruct runtime to throw the ambiguous resolution exception by means of disallowing
access to the target method and calling out to a throwing JIT helper.

I have expanded the original diamondshape test suite of positive and negative tests
for non-generic and generic variants to also include interfaces on struct and execution
of the SVM via ldftn and calli as well as the shared generic test requiring runtime lookup.

Thanks

Tomas

P.S. At this moment this change only targets CoreCLR. Mono and NativeAOT will need
additional work. Adding specific Crossgen2 support would be a possible future perf optimization.

trylek committed:
Move the diamondshape test under StaticVirtualMethods

trylek committed:
Modify the test to exercise static virtual methods

trylek committed:
Fixes to svm_diamondshape to make it build; initial runtime changes

trylek committed:
One more fix to the new diamondshape test; more runtime fixes

davidwrighton commented on src/tests/Loader/classloader/StaticVirtualMethods/DiamondShape/svm_diamondshape.cs:
```diff
@@ -0,0 +1,264 @@
+    
+    public static void Negative()
+    {
+        Console.WriteLine("Calling IFoo.Foo on Foo - expecting exception.");
```
We'll also need positive/negative cases for scenarios involving function pointers.

trylek replied:
I have copy & pasted and curated each of the test scenarios to exercise call / ldftn in the latest commit.

davidwrighton reviewed (changes requested):
I've identified some more work...

trylek committed:
Partial change for default interface implementations of SVMs

trylek committed:
Additional JIT changes to make default impl's of SVM resolve properly

trylek commented:
Hello @davidwrighton and @AndyAyersMS! I have managed to put together a batch of changes that make both the positive and negative diamond shape tests pass. I would very much welcome an initial review as much of the work is in JIT where I don't yet consider myself an expert. For now I plan that as the next step I'll work on the additional test scenarios per David's PR feedback (structs, ldftn).

Thanks a lot

Tomas

trylek commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -7633,14 +7648,30 @@ MethodTable::ResolveVirtualStaticMethod(MethodTable* pInterfaceType, MethodDesc*
                 }
+
+                // I have yet to figure out how to modify getCallInfo to support lazy throws upon
+                // ambiguous default interface method resolution. For now this effectively disables
```
I think I have managed to pull this off somehow but it remains to be seen whether my change is reasonable and / or whether it aligns with "established JIT practices". I would also very much welcome feedback regarding validation of the change - I think the good news is that the change basically doesn't alter any existing code paths, it just adds a couple of new ones, so I think that might reduce its riskiness.

trylek replied:
As explained in the PR description, based on our various discussions and based on debugging the code I have landed at a somewhat bifurcated implementation: when the SVM resolution is not possible at the time of getCallInfo, for shared generics we use the ConstrainedMethodEntrySlot dictionary entry to perform the resolution at runtime; for non-shared generics and for non-generic cases we use the accessAllowed flag per Andy's suggestion to disallow the call and instead call into a throwing JIT helper.

trylek commented:
<pre>
C:\git\runtime>C:\git\runtime\artifacts\tests\coreclr\windows.x64.Debug\Loader\classloader\StaticVirtualMethods\DiamondShape\svm_diamondshape_d\svm_diamondshape_d.cmd
BEGIN EXECUTION
 "C:\git\runtime\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe"   svm_diamondshape_d.dll
Calling IFoo.Foo on Foo - expecting exception.
Exception caught: System.Runtime.AmbiguousImplementationException: Ambiguous implementation found.
   at Program.Negative() in C:\git\runtime\src\tests\Loader\classloader\StaticVirtualMethods\DiamondShape\svm_diamondshape.il:line 566
Calling I1.Func on I47Class - expecting exception
Exception caught: System.Runtime.AmbiguousImplementationException: Ambiguous implementation found.
   at Program.Negative() in C:\git\runtime\src\tests\Loader\classloader\StaticVirtualMethods\DiamondShape\svm_diamondshape.il:line 600
Calling GI1<T>.Func on GI23Class<S> - expecting exception
Exception caught: System.Runtime.AmbiguousImplementationException: Ambiguous implementation found.
   at Program.Negative() in C:\git\runtime\src\tests\Loader\classloader\StaticVirtualMethods\DiamondShape\svm_diamondshape.il:line 635
Calling I1.Func on I4Class - expecting I4.Func
At I4.Func
PASS
Calling I1.Func on I8Class - expecting I8.Func
At I8.Func
PASS
Calling GI1.Func on GI4Class<object> - expecting GI4.Func<S>
System.Object, System.String, GI4
PASS
PASS
PASS
Expected: 100
Actual: 100
END EXECUTION - PASSED
PASSED
</pre>

trylek committed:
Fix temporary instrumentation

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5671,6 +5671,34 @@ HCIMPLEND_RAW
 EXTERN_C void JIT_ValidateIndirectCall();
 EXTERN_C void JIT_DispatchIndirectCall();
 
+HCIMPL2_VV(void*, JIT_DispatchVirtualStatic, MethodTable *constrainedType, MethodDesc *interfaceMethod)
```
The section of the file is for `INTEROP HELPERs`. This should be moved next to ` JIT_VirtualFunctionPointer`.

trylek replied:
Thanks for pointing that out, moved according to your suggestion.

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5671,6 +5671,34 @@ HCIMPLEND_RAW
 EXTERN_C void JIT_ValidateIndirectCall();
 EXTERN_C void JIT_DispatchIndirectCall();
 
+HCIMPL2_VV(void*, JIT_DispatchVirtualStatic, MethodTable *constrainedType, MethodDesc *interfaceMethod)
```
```suggestion
HCIMPL2(void*, JIT_DispatchVirtualStatic, MethodTable *constrainedType, MethodDesc *interfaceMethod)
```
`VV` is for methods that take arguments that do not fit into pointer size. Not the case here.

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5671,6 +5671,34 @@ HCIMPLEND_RAW
+
+    BOOL uniqueResolution;
+
+    FC_CAN_TRIGGER_GC();
```
There needs to be `HELPER_METHOD_FRAME` around the call to `ResolveVirtualStaticMethod`

trylek replied:
Thanks for pointing that out, fixed.

AndyAyersMS commented:
> I guess I would rather need a way to either make JIT emit a throw instead of the call into the caller

Not sure this is the best suggestion, but something similar already exists which keys off the `accessAllowed` field on `CORINFO_CALL_INFO`. If you set this to `CORINFO_ACCESS_ILLEGAL` you can then specify via `callsiteCalloutHelper` which helper the jit should call before making the call in question. So you can add a new helper that throws the proper exception.

If you then teach the jit that this new helper will always throw the jit will (eventually) remove the rest of what follows (that is, the actual call). It may be more challenging to try and suppress emitting the actual call as you'd need to mimic the effect of that call properly.

AndyAyersMS commented:
Also, jit changes look reasonable so far (modulo resolution of the invalid case).

trylek committed:
Address initial JanK's PR feedback

trylek commented:
For the <code>callsiteCalloutHelper</code>, I also don't see how we can propagate the typesystem objects (<code>MethodTable</code>, <code>MethodDesc</code>) to the error throw helpers to let us include them in the error message. This is not implemented today in the initial version of the JIT helper but I plan to add it as part of cleanup when removing the WIP tag and putting up a final version of the change.

jkotas commented:
> For the callsiteCalloutHelper, I also don't see how we can propagate the typesystem objects (MethodTable, MethodDesc) to the error throw helpers to let us include them in the error message. 

It is what `CORINFO_HELPER_DESC::args` are meant for.

trylek committed:
Add struct variants for the svm_diamondshape tests

This change extends the svm_diamondshape test to exercise value
types. As next step I'll work on testing the ldftn instruction.

Thanks

Tomas

trylek committed:
Add struct and ldftn testing per David's PR feedback

trylek commented:
I believe that by now I have addressed most of PR feedback. I have added JIT changes to cater for the negative tests per David's suggestion, I'm now trying to solicit feedback towards Andy's suggestion regarding simplification of some of these by leveraging <code>accessAllowed</code> / <code>callsiteCalloutHelper</code>. In accordance with David's suggestion I expanded the testing matrix by adding struct and ldftn variants to each of the negative and positive cases. I'm not yet clear on expanding the FCThrow statements in the new JIT helper to include the method and type in question as I don't yet fully understand where's the <code>kAmbiguousImplementationException</code> coming from, whether it expects particular arguments and how should I reflect that in the FCThrow statements.

trylek commented on src/tests/Loader/classloader/StaticVirtualMethods/DiamondShape/svm_diamondshape.cs:
```diff
@@ -0,0 +1,264 @@
+        T.Func<V>(out types);
+    }
+    
+    public static void Negative()
```
I have added struct counterparts to each of the test cases, thanks for pointing that out!

trylek commented on src/tests/Loader/classloader/StaticVirtualMethods/DiamondShape/svm_diamondshape.il:
```diff
@@ -0,0 +1,857 @@
+.class interface private abstract auto ansi I2
+       implements I1
+{
+  .method private hidebysig final 
```
<pre>
Unhandled exception. System.TypeLoadException: Method implementation on an interface 'IFoo2' from assembly 'svm_diamondshape, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' must be a final method.
</pre>

Am I right to understand this is an artifact of the pre-existing limitation that there can always be just one "actual" override? Should I strive to remove this limitation as it seems applicable no more?

davidwrighton replied:
This seems odd. These are static methods, and I don't believe final makes sense on static methods. I don't believe we need the static flag on implementation on the class either.

AlekseyTs replied:
> These are static methods, and I don't believe final makes sense on static methods.

I agree with this.

> I don't believe we need the static flag on implementation on the class either.

I am not sure I agree with this. The implemented method is static, so I think the implementing method should be marked static too. BTW, we do mark implementations in classes like that. Is there a reason. Here is what the spec says
```
II.23.1.10 Flags for methods [MethodAttributes]
Static 0x0010 Defined on type, else per instance  
```

AlekseyTs replied:
BTW, I think `final` should be used for re-abstraction. Such instance methods are marked "abstract virtual final" and static methods will be "static abstract virtual final". Do we have any examples for re-abstraction for static methods?

trylek replied:
@AlekseyTs - am I right to understand that by re-abstraction you mean something like a "newslot" in the sense of introducing a place in the type hierarchy where you'd suppress the SVM implementations defined "at a lower level" and require new ones to be supplied higher up the type hierarchy? If my interpretation is correct, I don't recall any targeted testing for this in our original SVM implementation but it shouldn't be rocket science to expand on the SVM diamond pattern test to include this case. I apologize if I misunderstood you, please clarify in such case.

AlekseyTs replied:
I'll respond with a code example because we are using many "previously undefined terms":
```
interface I1
{
    static abstract void M();
}

interface I2 : I1
{
    static void I1.M(){}
}

interface I3 : I2
{
    static abstract void I1.M(); // re-abstraction
}

class C2 : I2 {} // All is good

class C3: I3 {} // error CS0535: 'C3' does not implement interface member 'I1.M()'
```

It works the same way for instance methods today in the compiler. And it is supposed to work this way for instance methods in the runtime too.

trylek replied:
Thanks Aleksey, I believe your code example matches what I was trying to describe. I assume that the "diamond pattern" here may mean that in one edge of the type hierarchy graph the SVM has been "re-abstracted" while in another "the original SVM is implemented" thus forming another type of resolution ambiguity. We should certainly include some of these corner cases in unit testing for the feature.

AlekseyTs replied:
>  I believe your code example matches what I was trying to describe.

Great! It still looks like we need to get to a firm agreement around flags used in metadata for method declarations in each scenario.

trylek committed:
Remove 'final' check; add reabstraction test cases

trylek committed:
Fix a few syntax errors in the IL

SamMonoRT commented:
cc @BrzVlad

trylek committed:
Relax flag checking for reabstracted SVMs; runtime resolution test

trylek committed:
More changes based on David Wrighton's feedback

trylek committed:
Additional changes for shared generic support

trylek committed:
Fix calling convention flag setting in getCallInfo

trylek committed:
Fix test C# source code according to the IL; fix a few IL typos

trylek commented:
I believe I finally have a working implementation that passes all the negative and positive tests. I have also speculatively updated the C# test source code albeit it's not yet really used. I have removed the WIP tag and I believe the change is ready for a final round of reviews. Thanks everyone for your insightful comments and suggestions that turned out to be invaluable in implementing this change!

trylek commented:
/cc @BrzVlad

trylek committed:
Remove dummy 'else' block

trylek commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -6396,6 +6396,10 @@ MethodTableBuilder::PlaceMethodImpls()
         }
+        else
+        {
+            
```
I think it was just a leftover from previous experimental changes and needed deleting, removed in the latest commit.

MichalStrehovsky commented on src/coreclr/jit/importer.cpp:
```diff
@@ -9455,6 +9455,44 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
 
         switch (callInfo->kind)
         {
+            case CORINFO_VIRTUALSTATICCALL:
```
This looks almost like `CORINFO_CALL_CODE_POINTER` - can't we use that one?

trylek replied:
Thanks Michal for the suggestion. I originally tried that and hit some issues but now I tried to just switch it over and it works fine so I assume the problems I was hitting were unrelated and I changed it to <code>CORINFO_CALL_CODE_POINTER</code> per your suggestion in the latest commit.

trylek committed:
Fix failures in several pre-existing SVM tests

1. TryToResolveStaticVirtualMethodOnThisType doesn't need the
uniqueResolution flag - non-unique resolution on a single type
(i.o.w. multiple MethodImpl records resolving the same MethodDecl)
is always an error.

2. In getCallInfo, treat a SVM call as requiring runtime lookup
when the constrained type is a canonical subtype.

Thanks

Tomas

trylek committed:
Revert CORINFO_STATICVIRTUALCALL per Michal's PR feedback

As Michal pointed out, implementation of this call kind is quite
similar to CORINFO_CALL_CODE_POINTER; initially I had trouble
making it work but now the rest of the change is working fine
it turns out that CORINFO_CALL_CODE_POINTER works fine so I
removed the new kind I had originally introduced.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr outerloop

trylek committed:
Temporarily disable the SVM diamond shape test on mono with #70040

Disable the test before the Mono support is provided.

trylek commented:
OK, I now believe the change to be fully functional now - in my latest round of testing I only hit the two tailcall crashes on Windows arm in outerloop (#70042) and Mono crashes in the new SVM diamond shape test. I have created the issue #70040 to track the Mono implementation and temporarily blocked the test out in issues.targets. As next step we need to follow up regarding the Mono and NativeAOT implementations.

trylek committed:
Remove the issues.targets entry to see if Mono works now

trylek commented:
I have removed the Mono-specific issues.targets entry as it seems to be working now, I'll follow up based on test results.

AndyAyersMS reviewed (approved):
Happy to see the existing bail out logic for the jit worked out.  I approve the (now empty) set of jit changes.

jakobbotsch commented on src/coreclr/inc/corinfo.h:
```diff
@@ -646,6 +646,8 @@ enum CorInfoHelpFunc
     CORINFO_HELP_VALIDATE_INDIRECT_CALL,    // CFG: Validate function pointer
     CORINFO_HELP_DISPATCH_INDIRECT_CALL,    // CFG: Validate and dispatch to pointer
 
+    CORINFO_HELP_STATIC_VIRTUAL_AMBIGUOUS_RESOLUTION, // Throw AmbiguousResolutionException for failed static virtual method resolution
+
```
Sorry, just noticed this during a conflict resolution -- but this should get a matching change in CorInfoHelpFunc.cs, and I think it should also get a JIT-EE GUID update.

jakobbotsch replied:
If you plan to submit the JIT-EE GUID update, then can you also please take care of #69900 at the same time?

jakobbotsch replied:
In fact, it would be nice to name this `CORINFO_HELP_THROW_AMBIGUOUS_RESOLUTION_EXCEPTION` and move it up together with the other `CORINFO_HELP_THROW_*` helpers at the same time.
